README.rst
########
munchkin
########

Wordlist generator based on password cards

.. image:: https://img.shields.io/travis/gszathmari/munchkin.svg
    :target: https://travis-ci.org/gszathmari/munchkin
    :alt: Travis CI

.. image:: https://img.shields.io/codacy/3f42d3ee8060406d81e77dc6274bb671.svg
   :target: https://www.codacy.com/app/gszathmari/munchkin
   :alt: Codacy

.. image:: https://www.quantifiedcode.com/api/v1/project/c42903be20d644809c9c87c9b8c6b81f/badge.svg
   :target: https://www.quantifiedcode.com/app/project/c42903be20d644809c9c87c9b8c6b81f
   :alt: QuantifiedCode

.. image:: https://img.shields.io/pypi/dm/munchkin.svg
   :target: https://pypi.python.org/pypi/munchkin
   :alt: PyPI

.. image:: https://img.shields.io/requires/github/gszathmari/munchkin.svg
   :target: https://requires.io/github/gszathmari/munchkin/requirements/?branch=master
   :alt: Requirements Status

.. image:: https://img.shields.io/pypi/pyversions/munchkin.svg
   :alt: Python Versions

What is a password card?
========================

Password (aka. grid) card is a physical card that allows you to memorize your
passwords easier.

.. image:: https://raw.githubusercontent.com/gszathmari/munchkin/master/docs/images/password_card.png
  :alt: Password Card

The card contains a block of random alphanumeric characters. You just need to
choose and memorize a row, a column and a direction to pick a password from the
card.

Security
--------

Passwords cards heavily rely on physical security. The card should be kept in
your wallet or pocket, far away from watchful eyes around you.

In case the card is lost or recorded with a camera (e.g. CCTV), your password
is compromised therefore it should be changed immediately.

Features
========

Munchkin can generate wordlists from compromised password cards.

This utility generates potential passwords and dumps them onto the screen or
into a text file. The list of passwords can be used for brute-force attacks.

Supported Strategies
--------------------

Passwords from a password card can be chosen by the user based on different
strategies. He or she might read passwords from left to right, top to down or
diagonally.

Munchkin can generate passwords based on the most common card reading
strategies. The followings are samples only.

Left to Right
^^^^^^^^^^^^^

.. image:: https://raw.githubusercontent.com/gszathmari/munchkin/master/docs/images/left-to-right.png
   :alt: Left to Right

Top Down
^^^^^^^^

.. image:: https://raw.githubusercontent.com/gszathmari/munchkin/master/docs/images/top-down.png
   :alt: Top Down

Refer to the documentation for a comprehensive list of `password generating strategies`_.

.. _password generating strategies: https://github.com/gszathmari/munchkin/blob/master/docs/strategies.rst

Supported Cards
---------------

* Cards from http://passwordcard.org
* Any user-supplied custom card

Installing Munchkin
===================

The latest package is available on `PyPI`_ ::

  $ pip install munchkin

.. _PyPI: https://pypi.python.org/pypi/munchkin

Requirements
------------

This utility only runs on Python *2.6.x* and *2.7.x*

Usage Instructions
==================

The following section explains the basic usage of Munchkin. You can also use
the ``-h`` switch for getting more information on the individual features.

Operation Modes
---------------

There are two operation modes available. The first one generates cards similar
to the ones from http://passwordcard.org, and the second allows to bring your
own password cards.

passwordcard.org Cards
^^^^^^^^^^^^^^^^^^^^^^

Use the ``pcard`` selector to generate passwords from passwordcard.org cards ::

  $ munchkin pcard -h

This selector recognises the following options:

  -s str, --seed str  card number (e.g. *7eb3fbfa560d1d1e*)
  --symbols           include symbols
  --digits            incude digits

Custom Cards
^^^^^^^^^^^^

The ``custom`` selector allows to supply password cards by pasting them as a
block of text ::

  $ munchkin custom -h

This selector does not require any special options.

Settings
--------

The following switches are recognized for both card types

Password Length
^^^^^^^^^^^^^^^

Choose the minimum and maximum length of passwords to generate:

  --minlen num  minimum password length (*default: 6*)
  --maxlen num  maximum password length (*default: 12*)

Read Strategies
^^^^^^^^^^^^^^^

Select one or more strategies to generate passwords (refer
to `Supported Strategies`_ section for further explanation)

  -l, --left-to-right   Left to right
  -r, --right-to-left   Right to left
  -t, --top-down        Top left corner to bottom right
  -b, --bottom-up       Bottom right corner to top left

File Output
^^^^^^^^^^^

Dump passwords to a file instead of the terminal:

  -f name, --file name  Dump passwords to file

Examples
--------

Generate 6-8 digit passwords from a password card generated with seed the
initial seed of *7eb3fbfa560d1d1e* ::

  $ munchkin pcard -s 7eb3fbfa560d1d1e -l

Supply your own password card and generate passwords with multiple read
strategies ::

  $ munchkin custom -l -r -t -b

Links
=====

* `Source code on GitHub`_
* `Package on PyPI`_

.. _Source code on GitHub: https://github.com/gszathmari/munchkin
.. _Package on PyPI: https://pypi.python.org/pypi/munchkin

Contributors
============

* Gabor Szathmari - `@gszathmari`_

.. _@gszathmari: https://www.twitter.com/gszathmari

Credits
=======

* Python port of passwordcard.org algorithm: `olasd/passwordcard`_

.. _olasd/passwordcard: https://github.com/olasd/passwordcard


CHANGELOG.md
0.3.2 (2016-04-02)

- Bumping dependency version numbers

0.3.1 (2015-10-05)

- Fixing bug with version numbers

0.3.0 (2015-10-05)

- Add spiral card read strategy (resolves #3)
- Code refactoring around card reading strategies

0.2.7 (2015-09-30)

- Fixing bug related to reverse zig-zag
- Simplifying the handling of '-a' flag
- Refactoring strategies code into separate file

0.2.6 (2015-09-29)

- Restructure project directory
- Add Python version check
- Colorise console output
- Handle user interrupt at entering custom card

0.2.5 (2015-09-23)

- Minor code optimizations
- Adding QuantifiedCode build badge
- Adding requirements.io build badge
- Update imported package versions

0.2.4 (2015-09-23)

- Minor bugfixes and code optimizations

0.2.3 (2015-09-22)

- Fixing typo in console help

0.2.2 (2015-09-22)

- Fixing bug preventing from generating passwords in different ranges
- Add 'angled' card reading strategy (resolves #6)
- Add header of symbols if card is from passwordcard.org
- Modify build badges

0.2.1 (2015-09-22)

- Code refactoring

0.2.0 (2015-09-19)

- Fixing bug that prevented generating passwordcard.org cards with symbols (resolves #1)
- Adding zig-zag and reverse zig-zag algorithms (closes #2)
- Add diagonal read strategy (closes #4)
- Add '--all' switch to turn off each card read strategy (closes #7)
- Improve unit test coverage (closes #5)

0.1.3 (2015-09-18)

- Improve documentation
- Fixate numpy dependency version number

0.1.2 (2015-09-18)

- Remove universal flag from wheel package
- Improve documentation
- Add -v switch
- Add CONTRIBUTE.md file

0.1.1 (2015-09-18)

- Fix packaging issues

0.1.0 (2015-09-18)

- Beta release


munchkin/munchkin.py
# -*- coding: utf-8 -*-

import argparse
import random
import string
import subprocess
import sys
import os
import logging

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), 'lib'))

from colorama import Fore, Back, Style, init
from __about__ import __version__, __author__, __description__, __copyright__, __website__
from core.card import Card
from core.utils import supported_python_version

logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.INFO)

def password_dumper(args, card):
    """ Displays card and passwords or dumps them to file """
    # If this is a real terminal, display password card on screen
    if sys.stdout.isatty():
        logging.info(Fore.GREEN + "Printing password card" + Fore.RESET)
        print("\n%s" % card.print_card)
    # Dump passwords to file if this option was selected
    if args.file:
        f = args.file
        logging.info(Fore.CYAN + "Dumping to file: {0}".format(f.name) + Fore.RESET)
        for password in card.passwords:
            f.write("%s\n" % password)
        f.close()
        logging.info(Fore.GREEN + "Write completed" + Fore.RESET)
    # Just display the passwords if user pipes this command
    else:
        for password in card.passwords:
            print(password)

def generate_card_pcard(args, card):
    """ Generate card similar to ones on http://passwordcard.org """
    card.generate_password_card(symbols=args.symbols, digits=args.digits)
    password_dumper(args, card)

def generate_card_custom(args, card):
    """ Generate custom, user supplied password card """
    data = []
    print(Fore.YELLOW + Style.BRIGHT + "Please copy-paste password card below (press ENTER twice when done):\r\n")
    # Read card from terminal
    try:
        while True:
            input_str = raw_input()
            # Finish reading when user leaves line empty
            if input_str == "":
                break
            else:
                row = list(input_str)
                data.append(row)
    # Handle if user presses CTRL + C
    except KeyboardInterrupt:
        print("User interrupt")
        sys.exit(2)
    card.generate_custom_card(data)
    password_dumper(args, card)

def menu():
    """ Contains input arguments and launches controller """
    version_info = Fore.CYAN + "Munchkin wordlist generator version {0} by {1} {2}".format(__version__, __author__, __website__)

    parser = argparse.ArgumentParser()

    parser.add_argument('-v', '--version', help='version information', action='version', version=version_info)

    parent_parser = argparse.ArgumentParser(add_help=False)
    parent_parser.add_argument('-f', '--file', type=argparse.FileType('wb', 0), metavar='name', help='dump passwords to file')

    pwlen_options = parent_parser.add_argument_group('password lengths')
    pwlen_options.add_argument('--minlen', help='minimum password length (default: 6)', default=6, metavar='num')
    pwlen_options.add_argument('--maxlen', help='maximum password length (default: 12)', default=12, metavar='num')

    wordlist_options = parent_parser.add_argument_group('wordlist generator algorithms')
    wordlist_options.add_argument('-l', '--left-to-right', help='read card from top left to bottom right', action='store_true')
    wordlist_options.add_argument('-r', '--right-to-left', help='read card from bottom right to top left', action='store_true')
    wordlist_options.add_argument('-t', '--top-down', help='read card from top left to bottom left', action='store_true')
    wordlist_options.add_argument('-b', '--bottom-up', help='read card from bottom right to top right', action='store_true')
    wordlist_options.add_argument('-z', '--zig-zag', help='read card zig-zag from top left', action='store_true')
    wordlist_options.add_argument('-x', '--zig-zag-rev', help='read card zig-zag from bottom right', action='store_true')
    wordlist_options.add_argument('-d', '--diagonal', help='read card diagonally', action='store_true')
    wordlist_options.add_argument('-A', '--angled', help='read card at right-angle directions', action='store_true')
    wordlist_options.add_argument('-S', '--spiral', help='read card in spiral directions', action='store_true')
    wordlist_options.add_argument('-a', '--all', help='use each read strategy from above', action='store_true')

    subparsers = parser.add_subparsers(title='subcommands', description='valid subcommands', help='Generate wordlist based on cards from:')
    parser_pc = subparsers.add_parser('pcard', help="http://passwordcard.org", parents=[parent_parser])
    parser_custom = subparsers.add_parser('custom', help="your terminal", parents=[parent_parser])

    passwordcard_options = parser_pc.add_argument_group('passwordcard options')
    passwordcard_options.add_argument('-s', '--seed', help='card number (e.g. 7eb3fbfa560d1d1e)', required=True, metavar='str')
    passwordcard_options.add_argument('--symbols', help='include symbols', action='store_true')
    passwordcard_options.add_argument('--digits', help='incude digits', action='store_true')

    # Set controller functions
    passwordcard_options.set_defaults(func=generate_card_pcard)
    parser_custom.set_defaults(func=generate_card_custom)

    args = parser.parse_args()

    controller(args)

def controller(args):
    """ Verifies arguments and dispatches appropriate functions """
    # Verifies whether at least one strategy was selected
    if (args.left_to_right or
        args.right_to_left or
        args.top_down or
        args.bottom_up or
        args.zig_zag or
        args.zig_zag_rev or
        args.diagonal or
        args.angled or
        args.spiral or
        args.all) is False:
        logging.error("Please select at least one algorithm (hint: -l/-r/-t/-b/-z/-x/-d/-A/--all)\r\n")
        sys.exit(3)

    if args.all:
        args.left_to_right = True
        args.right_to_left = True
        args.top_down = True
        args.bottom_up = True
        args.zig_zag = True
        args.zig_zag_rev = True
        args.diagonal = True
        args.angled = True
        args.spiral = True

    # Create card and start processing
    card = Card(args)
    args.func(args, card)

# Displays banner and program version
def banner():
    """ Generates fancy banner """
    banner = """
 _______           _        _______           _       _________ _
(       )|\     /|( (    /|(  ____ \|\     /|| \    /\\__   __/( (    /|
| () () || )   ( ||  \  ( || (    \/| )   ( ||  \  / /   ) (   |  \  ( |
| || || || |   | ||   \ | || |      | (___) ||  (_/ /    | |   |   \ | |
| |(_)| || |   | || (\ \) || |      |  ___  ||   _ (     | |   | (\ \) |
| |   | || |   | || | \   || |      | (   ) ||  ( \ \    | |   | | \   |
| )   ( || (___) || )  \  || (____/\| )   ( ||  /  \ \___) (___| )  \  |
|/     \|(_______)|/    )_)(_______/|/     \||_/    \/\_______/|/    )_)\r\n"""
    print(Style.BRIGHT + "Munchkin {0} {1}".format(__version__, __description__))
    print(Fore.BLUE + "{0}".format(__copyright__))
    print(Style.DIM + banner)

def main(argv=None):
    """ Main program entry point """
    if supported_python_version() is False:
        print("Error: This utility only supports Python 2.6.x and 2.7.x")
        sys.exit(2)

    # Initialize colorama
    init(autoreset=True)

    # Only display the banner if the output is terminal
    if sys.stdout.isatty():
        banner()
    menu()
